### PR TITLE
[Inductor][MPS] Fix half-precision type mismatches in Metal shader codegen

### PR DESCRIPTION
**Pull request by mergennachin**
Metal Shading Language rejects implicit float-to-bfloat conversions, so
bare float literals like `0.0` in generated shaders cause compilation
failures when the target variable is `bfloat` (or `half`). Three codegen
methods were affected:

- `constant()` ignored its `dtype` parameter and returned raw literals.
- `masked()` assigned a bare literal in the else-branch (`} else tmp = 0.0;`).
- `where()` passed a bare literal through the ternary without casting.

All three now emit `static_cast<bfloat>(...)` / `static_cast<half>(...)`
where needed. Tests added for half-precision constants, reductions, and
conditionals.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo

### CODE REVIEW

File: test/inductor/test_mps_basic.py [hunk 1]
```diff
@@ -19,6 +19,7 @@
     [torch.bfloat16] if MACOS_VERSION < 14.0 else []
 )
 MPS_DTYPES = [t for t in get_all_dtypes() if t not in MPS_UNSUPPORTED_TYPES]
+MPS_HALF_DTYPES = [torch.float16] + ([torch.bfloat16] if MACOS_VERSION >= 14.0 else [])
```
Comment (malfet) [inline test/inductor/test_mps_basic.py]: We(And Apple) don't support MacOS-13 anymore. But I believe this list already exists
```suggestion
MPS_HALF_DTYPES = [torch.float16, torch.bfloat16]
```

File: test/inductor/test_mps_basic.py [hunk 2]
```diff
@@ -100,6 +101,39 @@ def foo(x):
...
+            return x + 1.0
+
+        self.common(
+            fn,
+            (make_tensor(1024, dtype=dtype, device=self.device),),
+            check_lowp=False,
+        )
+
+    @parametrize("dtype", MPS_HALF_DTYPES)
+    def test_half_masked(self, dtype):
```
Comment (malfet) [inline test/inductor/test_mps_basic.py]: Err, why does it called `test_half_masked`? when you call `sum`?

File: torch/_inductor/codegen/mps.py [hunk 3]
```diff
@@ -207,7 +207,14 @@ def to_dtype_bitcast(
 
     @staticmethod
     def constant(val: bool | float | int, dtype: torch.dtype) -> str:
-        return value_to_metal(val)
+        raw = value_to_metal(val)
+        if (
+            dtype in (torch.bfloat16, torch.float16)
```
Comment (malfet) [inline torch/_inductor/codegen/mps.py]: Do you know if this condition is ever hit? (I would have checked that one by adding an assert and running test suite)

But I would argue, that math op should return float values, and only downcast to low precision dtype when it's written out (as `opmath_t<bfloat> == float`)

File: torch/_inductor/codegen/mps.py [hunk 4]
```diff
@@ -244,12 +251,18 @@ def masked(mask: CSEVariable, body: sympy.Expr, other: CSEVariable) -> str:
             with V.kernel.compute.indent():
                 V.kernel.compute.splice(scoped_body)
                 V.kernel.compute.writeline(f"{var} = {rc};")
```
Comment (malfet) [inline torch/_inductor/codegen/mps.py]: Don't you need to do it here as well?

File: torch/_inductor/codegen/mps.py [hunk 5]
```diff
@@ -244,12 +251,18 @@ def masked(mask: CSEVariable, body: sympy.Expr, other: CSEVariable) -> str:
...
-            V.kernel.compute.writeline(f"}} else {var} = {other_str};")
+            V.kernel.compute.writeline(
+                f"}} else {var} = static_cast<{DTYPE_TO_METAL[rc.dtype]}>({other_str});"
```
Comment (malfet) [inline torch/_inductor/codegen/mps.py]: Nit (rc.dtype is often wrong/undefined)
```suggestion
                f"}} else {var} = static_cast<decltype(var)>({other_str});"
```

File: torch/_inductor/codegen/mps.py [hunk 6]
```diff
@@ -244,12 +251,18 @@ def masked(mask: CSEVariable, body: sympy.Expr, other: CSEVariable) -> str:
...
         return var
 
     @staticmethod
     def where(a: OpVarT, b: OpVarT, c: OpVarT) -> str:
-        return f"{a} ? {b} : {value_to_metal(c)}"
+        c_str = value_to_metal(c)
+        if isinstance(b, CSEVariable) and b.dtype in (torch.bfloat16, torch.float16):
+            assert b.dtype is not None
+            c_str = f"static_cast<{DTYPE_TO_METAL[b.dtype]}>({c_str})"
+        return f"{a} ? {b} : {c_str}"
```
Comment (malfet) [inline torch/_inductor/codegen/mps.py]: Nit (i.e. always leave it to compiler rather than codegen)
```suggestion
        return f"{a} ? {b} : static_cast<decltype({a})>({value_to_metal(c)})"
```

File: torch/_inductor/codegen/mps.py (hunk 3 above)
Comment (mergennachin) [reply torch/_inductor/codegen/mps.py]: Yes, the condition is hit. Verified empirically x + 1.0 with bf16 input calls constant(val=1.0, dtype=torch.bfloat16). Same for x * 2, x - 0.5, and torch.where with scalar constants.


You're right that math ops should return float values. The current fix does a pointless round-trip: the literal is float, we cast it to bfloat, then Metal implicitly promotes it back to float for arithmetic (since loads are already upcast to float32). I'll drop the cast in constant() and just return the raw literal. It's already float-compatible and matches the float32 compute dtype established by load().

File: test/inductor/test_torchinductor.py [hunk 7]
```diff
@@ -15838,6 +15838,36 @@ def run_session(x_param, y_param, size, device):
...
 
+    def test_half_constant(self):
+        for dtype in [torch.float16, torch.bfloat16]:
+            if not self.is_dtype_supported(dtype):
+                continue
+            self.common(
+                lambda x: x + 1.0,
+                (make_tensor(1024, dtype=dtype, device=self.device),),
+                check_lowp=False,
+            )
```
Comment (malfet) [inline test/inductor/test_torchinductor.py]: Nit (check_lowp indeed checks for torch.half)
```suggestion
    def test_bfloat_constant(self):
        if not self.is_dtype_supported(torch.bfloat16):
            continue
        self.common(
            lambda x: x + 1.0,
            (make_tensor(1024, dtype=torch.bfloat16, device=self.device),),
        )
```

File: test/inductor/test_torchinductor.py [hunk 8]
```diff
@@ -15838,6 +15838,36 @@ def run_session(x_param, y_param, size, device):
...
+
+    def test_half_reduction(self):
+        for dtype in [torch.float16, torch.bfloat16]:
+            if not self.is_dtype_supported(dtype):
+                continue
+            self.common(
+                lambda x: x.sum(),
+                (make_tensor(1024, dtype=dtype, device=self.device),),
+                check_lowp=False,
+            )
```
Comment (malfet) [inline test/inductor/test_torchinductor.py]: Alternatively use paramterize 
```suggestion
    @paramtertize(dtype, [torch.float16, torch.bfloat16])
    def test_lowp_reduction(self, dtype):
            if not self.is_dtype_supported(dtype):
                continue
            self.common(
                lambda x: x.sum(),
                (make_tensor(1024, dtype=dtype, device=self.device),),
                check_lowp=False,
            )
```

**Review by malfet — APPROVED** [review]
